### PR TITLE
node grpc spike dashboard to server

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at c251d76</samp>

This pull request adds new files that define and generate code for a dummy service for reliability testing using various protobuf and connect libraries. The files include a protobuf file, go files for grpc and connect-go clients and servers, and typescript files for connect-web clients and messages.

</details>

## Related Issue(s)

Fixes EXP-571

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - ak-node-grpc</li>
	<li><b>🔗 URL</b> - <a href="https://ak-node-grpc.preview.gitpod-dev.com/workspaces" target="_blank">ak-node-grpc.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - ak-node_grpc-gha.16710</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-ak-node-grpc%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/session-handler.ts [hunk 1]
```diff
@@ -90,38 +91,37 @@ export class SessionHandler {
...
-            // We set the user object on the request to signal the user is authenticated.
-            // Passport uses the `user` property on the request to determine if the session
-            // is authenticated.
-            req.user = user;
-
-            // Trigger the next middleware in the chain.
-            next();
+            return await this.userService.findUserById(subject, subject);
         } catch (err) {
             log.warn("Failed to authenticate user with JWT Session", err);
```
Comment (akosyakov) [inline components/server/src/session-handler.ts:123]: not sure here, maybe we should call next with err, but maybe not

File: components/gitpod-protocol/src/experiments/types.ts [hunk 2]
```diff
@@ -28,6 +28,9 @@ export interface Attributes {
 
     // Host name of the Gitpod installation (mapped to "custom.gitpod_host")
     gitpodHost?: string;
+
+    // The type of the gRPC client request (mapped to "custom.grpc_type")
+    grpcType?: "unary" | "server-stream";
```
Comment (akosyakov) [inline components/gitpod-protocol/src/experiments/types.ts:33]: let's remove it

File: components/dashboard/src/service/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+import { GitpodHostUrl } from "@gitpod/gitpod-protocol/lib/util/gitpod-host-url";
+import { MetricsReporter } from "@gitpod/public-api/lib/metrics";
+
+const originalConsoleError = console.error;
+
+const options = {
+    gitpodUrl: new GitpodHostUrl(window.location.href).withoutWorkspacePrefix().toString(),
+    clientName: "dashboard",
+    clientVersion: "",
```
Comment (akosyakov) [inline components/dashboard/src/service/metrics.ts:15]: feature flag it

File: components/public-api/typescript/src/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+        const data = {
+            name,
+            labels,
+            value,
+        };
+        const resp = await fetch(`https://${this.metricsHost}/metrics-api/metrics/counter/add/${name}`, {
+            method: "POST",
+            headers: {
+                "Content-Type": "application/json",
+                "X-Client": this.options.clientName,
```
Comment (akosyakov) [inline components/public-api/typescript/src/metrics.ts:342]: metric_client is "unknown" for some reason

File: components/server/src/session-handler.ts (hunk 1 above)
Comment (akosyakov) [reply components/server/src/session-handler.ts:123]: This PR did not change the semantic though. Pleas review for it 🙏

File: components/public-api/typescript/src/metrics.ts [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+            if (settled) {
+                stopTimer({ grpc_code: status ? Code[status] : "OK" });
+                GRPCMetrics.handled({ ...labels, code: status ? Code[status] : "OK" });
+            }
+        }
+    };
+}
+
+export class MetricsReporter {
+    private static readonly REPORT_INTERVAL = 10000;
```
Comment (akosyakov) [inline components/public-api/typescript/src/metrics.ts:247]: we can increase it to 1 min and instead do a final flush with https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon if it turns out to be a problem later

**Comment by geropl** [comment]
Starting to review now.

File: components/dashboard/src/service/service.tsx [hunk 6]
```diff
@@ -56,7 +58,67 @@ export function getGitpodService(): GitpodService {
         const service = _gp.gitpodService || (_gp.gitpodService = require("./service-mock").gitpodServiceMock);
         return service;
     }
-    const service = _gp.gitpodService || (_gp.gitpodService = createGitpodService());
+    let service = _gp.gitpodService;
+    if (!service) {
+        let user: any;
```
Comment (geropl) [inline components/dashboard/src/service/service.tsx:63]: nit: Would be cool to encapsulate the test code here a bit more. But as along as we keep track of removing it later it's fine with me. :+1:

File: components/server/src/api/server.ts [hunk 7]
```diff
@@ -50,4 +73,120 @@ export class API {
...
+        app.use(
+            expressConnectMiddleware({
+                routes: (router: ConnectRouter) => {
+                    for (const [type, impl] of [service(HelloService, this.apiHelloService)]) {
+                        router.service(type, new Proxy(impl, this.interceptService(type)));
+                    }
+                },
+            }),
+        );
+        // TODO(al) cover unhandled cases
```
Comment (geropl) [inline components/server/src/api/server.ts:87]: :+1:

File: components/server/src/api/server.ts [hunk 8]
```diff
@@ -50,4 +73,120 @@ export class API {
...
+                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                    deferred.promise.then((err) => {
+                        const grpc_code = err ? Code[err.code] : "OK";
+                        grpcServerHandled.labels(grpc_service, grpc_method, grpc_type, grpc_code).inc();
+                        stopTimer({ grpc_code });
+                    });
+
+                    const context = args[1] as HandlerContext;
+                    async function call<T>(): Promise<T> {
+                        const user = await self.verify(context);
```
Comment (geropl) [inline components/server/src/api/server.ts:139]: nit: As discussed IMO we could pull this (and some of the metrics here) into middleware. But not a priority for this PR.

File: components/server/src/container-module.ts [hunk 9]
```diff
@@ -332,11 +328,7 @@ export const productionContainerModule = new ContainerModule(
         bind(RelationshipUpdater).toSelf().inSingletonScope();
 
         // grpc / Connect API
-        bind(APIUserService).toSelf().inSingletonScope();
-        bind(APITeamsService).toSelf().inSingletonScope();
-        bind(APIWorkspacesService).toSelf().inSingletonScope();
-        bind(APIStatsService).toSelf().inSingletonScope();
-        bind(API).toSelf().inSingletonScope();
+        API.contribute(bind);
```
Comment (geropl) [inline components/server/src/container-module.ts:331]: So far we often called these methods `abcBinder` or `bindXyz`, bot contribute works for me.

File: components/dashboard/src/service/metrics.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import { getExperimentsClient } from "../experiments/client";
+
+const originalConsoleError = console.error;
+
+const options = {
+    gitpodUrl: new GitpodHostUrl(window.location.href).withoutWorkspacePrefix().toString(),
+    clientName: "dashboard",
+    clientVersion: "",
+    logError: originalConsoleError.bind(console),
+    isEnabled: () => getExperimentsClient().getValueAsync("dashboard_metrics_enabled", false, {}),
```
Comment (geropl) [inline components/dashboard/src/service/metrics.ts:18]: @akosyakov We could default to `true` here I think, to favor "more data" over "be totally safe against potential issues".

Comment (geropl) [reply components/dashboard/src/service/metrics.ts:18]: Ah, did not think about Dedicated. Yeah, I guess then false is better.

**Comment by geropl** [comment]
For some reason I never see any Hello RPC method triggered... metrics work fine though. :thinking: I wonder whether that feature flag evaluates to true....

File: components/dashboard/src/service/service.tsx [hunk 11]
```diff
@@ -56,7 +58,67 @@ export function getGitpodService(): GitpodService {
...
+                    if (propKey === "getWorkspace") {
+                        try {
+                            return await target[propKey](...args);
+                        } finally {
+                            const grpcType = "unary";
+                            // emulates frequent unary calls to public API
+                            const isTest = await getExperimentsClient().getValueAsync(
+                                "public_api_dummy_reliability_test",
+                                false,
+                                { user },
```
Comment (geropl) [inline components/dashboard/src/service/service.tsx:81]: Can we add this (same as here):
```suggestion
                                { user, gitpodHost: window.location.host },
```
I thought we would inject this in the proxy route, but it seems we do it in the query.

File: components/dashboard/src/service/service.tsx [hunk 12]
```diff
@@ -56,7 +58,67 @@ export function getGitpodService(): GitpodService {
...
+                    return target[propKey](...args);
+                };
+            },
+        });
+        (async () => {
+            const grpcType = "server-stream";
+            // emulates server side streaming with public API
+            while (true) {
+                const isTest = await getExperimentsClient().getValueAsync("public_api_dummy_reliability_test", false, {
+                    user,
```
Comment (geropl) [inline components/dashboard/src/service/service.tsx:103]:
```suggestion
                    user, gitpodHost: window.location.host
```
Same here.

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works. :heavy_check_mark: 

Was just tired yesterday :sweat_smile: 

Two small nits: https://github.com/gitpod-io/gitpod/pull/18691/files#r1326861983

**Pull request merged**